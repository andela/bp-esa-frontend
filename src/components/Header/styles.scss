#header *, *::before, *::after{
  padding: 0;
  margin: 0;
}

#header {
  background-color: rgb(51, 89, 223);
  display: inline-grid;
  padding: 10px 20px;
  width: 100%;
  box-sizing: border-box;
  font-size: 16px;

  place-items: center;
  grid-template-columns: repeat(2, 50%);
  .brand {
    display: inline-grid;
    place-items: center;
    justify-self: left;

    a {
      text-decoration: none;

      display: inline-grid;
      place-items: center;
      grid-template-columns: repeat(2, min-content);
      grid-column-gap: 0.7em;
      justify-self: left;
      .header-logo {
        position: relative;
        img {
          width: 35px;
        }

        &::after {
          position: absolute;
          content: 'A';
          color: #FFF;
          font-size: 20px;
          top: 0;
          left: 0;
          width: 100%;
          height: 100%;
          display: inline-grid;
          place-content: center;
          font-weight: lighter;
          font-family: Arial;
        }
      }
      &::after {
        display: block;
        content: 'ESA';
        color: #FBFBFB;
        user-select: text;
        font-size: 1.25em;
        font-weight: 600;
      }
      .text {
        display: none;
        font-weight: 600;
      }
    }

  }

  .user-info-container{
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    padding: 0.2375em 0.4375em;
    border: 1px solid #FBFBFB;
    border-radius: 5px;

    display: inline-grid;
    place-items: center;
    grid-template-columns: repeat(2, min-content);
    grid-column-gap: 8px;
    justify-self: right;

    &:hover{
      cursor: pointer;
    }

    .current-user {
      display: inline-grid;
      place-items: center;
      grid-template-columns: repeat(2, min-content);
      grid-column-gap: 10px;
      .user-name {
        font-size: 1em;
        &::before {
          display: block;
          content: attr(data-initials);
          color: #f4f4f4;
          line-height: 17px;
          font-weight: 600;
        }

        &::after {
          display: none;
          content: attr(data-name);
          color: #f4f4f4;
          font-weight: 600;
        }
      }
      .image-container {
        color: white;
        .user-image {
          border-radius: 50%;
          height: 25px;
          width: 25px;
        }
      }
    }
    .caret-down {
      color: white;
      padding: 0 5px;
      position: relative;

      i.fa-caret-down {
        &::before {
          display: block;
          transform: rotate(0);
          transition: transform 250ms;
        }
      }

      &.visible {
        i.fa-caret-down {
          &::before {
            display: block;
            transform: rotate(180deg);
          }
        }
      }

      .signout-dropdown-parent {
        width: 100px;
        height: 35px;
        position: absolute;
        right: -12px;
        margin: 7px 0 0 0;
        overflow: hidden;
        user-select: none;
        z-index: -10;

        &.visible {
          z-index: 10;
        }
        
        .signout-dropdown {
          font-size: 0.7em;
          position: absolute;
          top: -50px;
          right: 4px;
          width: max-content;
          height: max-content;
          padding: 0.4em 0.3em;
          background-color: white;
          border-radius: 2px;

          transition: top ease-out 300ms;
          &.visible {
            box-shadow: 2px 2px 2px 1px rgba(0, 0, 0, 0.3);
            top: 0;
          }
          .logout-button {
            border: none;
            border-radius: 2px;

            cursor: pointer;
            color: #667;
            width: fit-content;
            padding: 0.2em 1.2em;

            i {
              margin-right: 3px;
            }
            &:hover{
              background-color: #f0f0f0f0;
            }
          }
        }
      }

    }
  }
}

@media only screen and (min-width: 450px) {
  #header {
    padding: 10px 30px;
    .brand {
      a {
        grid-template-columns: repeat(2, max-content);
        .logo {
          width: 35px;
        }
        .text {
          color: #FBFBFB;
          font-size: 1.25em;
          line-height: 1.5em;
          display: block;
        }

        &::after {
          display: none;
        }
      }

    }

    .user-info-container{
      grid-column-gap: 3px;
      .current-user {
        grid-template-columns: repeat(2, max-content);
        grid-column-gap: 10px;
        .user-name {
          &::before {
            display: none;
          }

          &::after {
            display: block;
            line-height: 1.125em;
          }
        }
        .image-container {
          color: white;
          .user-image {
            border-radius: 50%;
            height: 25px;
            width: 25px;
          }
        }
      }
      .caret-down {
        .signout-dropdown-parent {
          width: 150px;
          height: 60px;
          .signout-dropdown {
            font-size: 1em;
            top: -50px;
          }
        }
      }
    }
  }
}
