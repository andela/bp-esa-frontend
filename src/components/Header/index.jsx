import React, { PureComponent } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { notify } from 'react-notify-toast';
import { doSignOut } from '../../firebase';
import './styles.scss';

class Header extends PureComponent {
  state = {
    signoutDropDownIsVisible: false,
  }

  toggleSignoutDropDown = (event) => {
    event.preventDefault();
    const { signoutDropDownIsVisible } = this.state;
    this.setState({
      signoutDropDownIsVisible: !signoutDropDownIsVisible,
    });
  }

  onLogout = () => {
    const { removeCurrentUser, history } = this.props;
    doSignOut().then((error) => {
      if (error) {
        notify.show('Unable to logout!', 'error');
      } else {
        removeCurrentUser();
        history.push('/login');
        notify.show('You have Logged out Successfully!', 'success');
      }
    });
  }

  render() {
    const { signoutDropDownIsVisible } = this.state;
    const { currentUser } = this.props;
    const { additionalUserInfo: { profile: { name, picture } } } = currentUser;
    return (
      <div id="header">
        <div className="brand">
          {/* eslint-disable jsx-a11y/anchor-is-valid */}
          <a href="#">
            <img className="logo" src="/logo.png" alt="Andela Logo" />
            <span className="text">ESA Reporting</span>
          </a>
        </div>
        <div className="signout">
          <span className="user-name">
            Hello,&nbsp;
            { name }
            !
          </span>
          <a href="#" onClick={this.toggleSignoutDropDown}>
            { picture
              ? <img src={picture} alt="user-icon" className="user-image" />
              : <i className="fa fa-2x fa-user-circle" />
            }
            <i className="fa fa-caret-down custom" />
          </a>
          <div
            className={classNames('signout-dropdown', { visible: signoutDropDownIsVisible })}
          >
            <a id="sign-out" href="#" onClick={() => this.onLogout()}>
              <i className="fa fa-sign-out-alt" />&nbsp;
              <span>Sign Out</span>
            </a>
          </div>
        </div>
      </div>
    );
  }
}

Header.propTypes = {
  currentUser: PropTypes.object.isRequired,
  removeCurrentUser: PropTypes.func,
  history: PropTypes.object.isRequired,
};

Header.defaultProps = {
  removeCurrentUser: () => {},
};

export default Header;
