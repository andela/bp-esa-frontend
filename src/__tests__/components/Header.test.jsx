import React from 'react';
import { notify } from 'react-notify-toast';
import { doSignOut } from '../../firebase';
import Header from '../../components/Header';

jest.mock('../../firebase');
const props = {
  currentUser: {
    UserInfo: {
      firstName: 'Jane',
      lastName: 'Doe',
      picture: 'https://test-image.jpg',
    },
  },
  removeCurrentUser: jest.fn(),
  history: {
    push: jest.fn(),
  },
};
let wrapper;

const getComponent = () => {
  if (!wrapper) {
    wrapper = shallow(<Header {...props} />);
  }
  return wrapper;
};

it('should render as expected', () => {
  const component = getComponent();
  expect(component).toMatchSnapshot();
});

describe('onLogout() method', () => {
  const renderedComponent = mount(<Header {...props} />);
  Object.defineProperty(notify, 'show', { value: () => jest.fn(), writable: true });
  it('should call onLogout()', () => {
    const spy = jest.spyOn(renderedComponent.instance(), 'onLogout');
    renderedComponent.find('.logout-button').simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('should call onLogout() error', async () => {
    await doSignOut.mockImplementationOnce(
      () => new Promise((resolve) => {
        resolve('Some error');
      }),
    );
    const header = new Header(props);
    const notifyError = await header.onLogout();
    expect(typeof notifyError === 'function').toEqual(true);
  });
});

describe('toggleSignoutDropDown() method', () => {
  it('should call toggleSignoutDropDown()', () => {
    const event = { preventDefault: () => {} };
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'toggleSignoutDropDown');
    renderedComponent.toggleSignoutDropDown(event);
    expect(renderedComponent.toggleSignoutDropDown.calledOnce).toEqual(true);
    expect(getComponent().state().signoutDropDownIsVisible).toEqual(true);
  });
});
