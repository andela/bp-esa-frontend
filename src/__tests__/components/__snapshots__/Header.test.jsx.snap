// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    currentUser={
      Object {
        "UserInfo": Object {
          "firstName": "Jane",
          "lastName": "Doe",
          "picture": "https://test-image.jpg",
        },
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    removeCurrentUser={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="brand"
        >
          <a
            href="/"
          >
            <img
              alt="Andela Logo"
              className="logo"
              src="/logo.png"
            />
            <span
              className="text"
            >
              ESA Reporting
            </span>
          </a>
        </div>,
        <div
          className="signout"
        >
          <div
            className="user-name"
          >
            Hello, 
            Jane Doe
            !
          </div>
          <div
            className="image-container"
            onClick={[Function]}
          >
            <img
              alt="user-icon"
              className="user-image"
              src="https://test-image.jpg"
            />
            <i
              className="fa fa-caret-down custom"
            />
            <div
              className="signout-dropdown"
            >
              <button
                className="logout-button"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fa fa-sign-out-alt"
                />
                 
                <span>
                  Sign Out
                </span>
              </button>
            </div>
          </div>
        </div>,
      ],
      "id": "header",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <a
            href="/"
          >
            <img
              alt="Andela Logo"
              className="logo"
              src="/logo.png"
            />
            <span
              className="text"
            >
              ESA Reporting
            </span>
          </a>,
          "className": "brand",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="Andela Logo"
                className="logo"
                src="/logo.png"
              />,
              <span
                className="text"
              >
                ESA Reporting
              </span>,
            ],
            "href": "/",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "Andela Logo",
                "className": "logo",
                "src": "/logo.png",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "ESA Reporting",
                "className": "text",
              },
              "ref": null,
              "rendered": "ESA Reporting",
              "type": "span",
            },
          ],
          "type": "a",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="user-name"
            >
              Hello, 
              Jane Doe
              !
            </div>,
            <div
              className="image-container"
              onClick={[Function]}
            >
              <img
                alt="user-icon"
                className="user-image"
                src="https://test-image.jpg"
              />
              <i
                className="fa fa-caret-down custom"
              />
              <div
                className="signout-dropdown"
              >
                <button
                  className="logout-button"
                  onClick={[Function]}
                  type="button"
                >
                  <i
                    className="fa fa-sign-out-alt"
                  />
                   
                  <span>
                    Sign Out
                  </span>
                </button>
              </div>
            </div>,
          ],
          "className": "signout",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Hello, ",
                "Jane Doe",
                "!",
              ],
              "className": "user-name",
            },
            "ref": null,
            "rendered": Array [
              "Hello, ",
              "Jane Doe",
              "!",
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <img
                  alt="user-icon"
                  className="user-image"
                  src="https://test-image.jpg"
                />,
                <i
                  className="fa fa-caret-down custom"
                />,
                <div
                  className="signout-dropdown"
                >
                  <button
                    className="logout-button"
                    onClick={[Function]}
                    type="button"
                  >
                    <i
                      className="fa fa-sign-out-alt"
                    />
                     
                    <span>
                      Sign Out
                    </span>
                  </button>
                </div>,
              ],
              "className": "image-container",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "user-icon",
                  "className": "user-image",
                  "src": "https://test-image.jpg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-caret-down custom",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <button
                    className="logout-button"
                    onClick={[Function]}
                    type="button"
                  >
                    <i
                      className="fa fa-sign-out-alt"
                    />
                     
                    <span>
                      Sign Out
                    </span>
                  </button>,
                  "className": "signout-dropdown",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <i
                        className="fa fa-sign-out-alt"
                      />,
                      " ",
                      <span>
                        Sign Out
                      </span>,
                    ],
                    "className": "logout-button",
                    "onClick": [Function],
                    "type": "button",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "className": "fa fa-sign-out-alt",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "i",
                    },
                    " ",
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Sign Out",
                      },
                      "ref": null,
                      "rendered": "Sign Out",
                      "type": "span",
                    },
                  ],
                  "type": "button",
                },
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="brand"
          >
            <a
              href="/"
            >
              <img
                alt="Andela Logo"
                className="logo"
                src="/logo.png"
              />
              <span
                className="text"
              >
                ESA Reporting
              </span>
            </a>
          </div>,
          <div
            className="signout"
          >
            <div
              className="user-name"
            >
              Hello, 
              Jane Doe
              !
            </div>
            <div
              className="image-container"
              onClick={[Function]}
            >
              <img
                alt="user-icon"
                className="user-image"
                src="https://test-image.jpg"
              />
              <i
                className="fa fa-caret-down custom"
              />
              <div
                className="signout-dropdown"
              >
                <button
                  className="logout-button"
                  onClick={[Function]}
                  type="button"
                >
                  <i
                    className="fa fa-sign-out-alt"
                  />
                   
                  <span>
                    Sign Out
                  </span>
                </button>
              </div>
            </div>
          </div>,
        ],
        "id": "header",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <a
              href="/"
            >
              <img
                alt="Andela Logo"
                className="logo"
                src="/logo.png"
              />
              <span
                className="text"
              >
                ESA Reporting
              </span>
            </a>,
            "className": "brand",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <img
                  alt="Andela Logo"
                  className="logo"
                  src="/logo.png"
                />,
                <span
                  className="text"
                >
                  ESA Reporting
                </span>,
              ],
              "href": "/",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "Andela Logo",
                  "className": "logo",
                  "src": "/logo.png",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "ESA Reporting",
                  "className": "text",
                },
                "ref": null,
                "rendered": "ESA Reporting",
                "type": "span",
              },
            ],
            "type": "a",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="user-name"
              >
                Hello, 
                Jane Doe
                !
              </div>,
              <div
                className="image-container"
                onClick={[Function]}
              >
                <img
                  alt="user-icon"
                  className="user-image"
                  src="https://test-image.jpg"
                />
                <i
                  className="fa fa-caret-down custom"
                />
                <div
                  className="signout-dropdown"
                >
                  <button
                    className="logout-button"
                    onClick={[Function]}
                    type="button"
                  >
                    <i
                      className="fa fa-sign-out-alt"
                    />
                     
                    <span>
                      Sign Out
                    </span>
                  </button>
                </div>
              </div>,
            ],
            "className": "signout",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Hello, ",
                  "Jane Doe",
                  "!",
                ],
                "className": "user-name",
              },
              "ref": null,
              "rendered": Array [
                "Hello, ",
                "Jane Doe",
                "!",
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <img
                    alt="user-icon"
                    className="user-image"
                    src="https://test-image.jpg"
                  />,
                  <i
                    className="fa fa-caret-down custom"
                  />,
                  <div
                    className="signout-dropdown"
                  >
                    <button
                      className="logout-button"
                      onClick={[Function]}
                      type="button"
                    >
                      <i
                        className="fa fa-sign-out-alt"
                      />
                       
                      <span>
                        Sign Out
                      </span>
                    </button>
                  </div>,
                ],
                "className": "image-container",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "user-icon",
                    "className": "user-image",
                    "src": "https://test-image.jpg",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "fa fa-caret-down custom",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <button
                      className="logout-button"
                      onClick={[Function]}
                      type="button"
                    >
                      <i
                        className="fa fa-sign-out-alt"
                      />
                       
                      <span>
                        Sign Out
                      </span>
                    </button>,
                    "className": "signout-dropdown",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        <i
                          className="fa fa-sign-out-alt"
                        />,
                        " ",
                        <span>
                          Sign Out
                        </span>,
                      ],
                      "className": "logout-button",
                      "onClick": [Function],
                      "type": "button",
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "className": "fa fa-sign-out-alt",
                        },
                        "ref": null,
                        "rendered": null,
                        "type": "i",
                      },
                      " ",
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Sign Out",
                        },
                        "ref": null,
                        "rendered": "Sign Out",
                        "type": "span",
                      },
                    ],
                    "type": "button",
                  },
                  "type": "div",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
